File: launcher/minimal/src/main/java/ink/reactor/launcher/console/JLineConsole.java
package ink.reactor.launcher.console;

import ink.reactor.launcher.ReactorLauncher;
import lombok.experimental.UtilityClass;
import org.jline.reader.LineReader;
import org.jline.reader.LineReaderBuilder;
import org.jline.terminal.Terminal;
import org.jline.terminal.TerminalBuilder;

import java.io.IOException;

@UtilityClass
public final class JLineConsole {

    public static Console createConsole() {
        final Terminal terminal;
        try {
            terminal = TerminalBuilder.builder()
                .system(true)
                .build();
        } catch (final IOException e) {
            System.err.println("Error on creating terminal");
            e.printStackTrace(System.err);
            return null;
        }

        final LineReader reader = LineReaderBuilder.builder()
            .terminal(terminal)
            .highlighter(new ConsoleHighlighter())
            .completer(new ConsoleCompleter())
            .build();

        final Console console = new Console(terminal, reader);
        ReactorLauncher.STOP_TASKS.add(console::stop);

        return console;
    }
}

File: launcher/minimal/src/main/kotlin/ink/reactor/launcher/console/JLineConsole.kt
package ink.reactor.launcher.console

import ink.reactor.kernel.Reactor
import org.jline.reader.LineReaderBuilder
import org.jline.terminal.Terminal
import org.jline.terminal.TerminalBuilder
import java.io.IOException

object JLineConsole {
    fun createConsole(): Console? {
        val terminal: Terminal
        try {
            terminal = TerminalBuilder.builder()
                .system(true)
                .build()
        } catch (e: IOException) {
            System.err.println("Error on creating terminal")
            e.printStackTrace(System.err)
            return null
        }

        val reader = LineReaderBuilder.builder()
            .terminal(terminal)
            .highlighter(ConsoleHighlighter())
            .completer(ConsoleCompleter())
            .build()

        val console = Console(terminal, reader)
        Reactor.addStopTask{console.stop()}

        return console
    }
}
